main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala

```diff
@@ -730,4 +730,236 @@ object SimplifiedAst {
 
   case class Property(law: Law, exp: SimplifiedAst.Expression, loc: SourceLocation) extends SimplifiedAst
 
+
+  /**
+   * A trait to consume a SimplifiedAst.Expression
+   * and produce a new Expression that's modified
+   * in some way. Typical usage is to extend the
+   * trait and override which methods corresponding
+   * to whatever case of Expression you're interested
+   * in and then call the `foldExpression` method
+   * to begin modifying a given expression.
+   */
+  trait ExpressionFolder {
```

Comment:
I am not super convinced by the usefulness of the visitor pattern, but let us see how it plays out.